book/src/design/circuit/gadgets/ecc.md

```diff
@@ -1 +1,76 @@
-# Elliptic Curve Cryptography
\ No newline at end of file
+# Elliptic Curve Cryptography
+
+## Incomplete addition
+Inputs: $P = (x_P, y_P), Q = (x_Q, y_Q)$
+Output: $A = P + Q = (x_A, y_A)$
+
+Formulae:
+- $\lambda \cdot (x_p - x_{q}) = y_p - y_{q}$
+- $x_{a} = \lambda^2 - x_{q} - x_p$
+- $y_{a} = \lambda(x_{q} - x_{a}) - y_{q}$
+
+Substituting for $\lambda$, we get the constraints:
+- $(x_{a} + x_{q} + x_p) \cdot (x_p - x_q)^2 - (y_p - y_{q})^2 = 0$
```

Comment:
This constraint is unsatisfiable for `P + (-P)`, and so we can't use it with arbitrary inputs (for the performance optimisation where ⊥ is represented either by unconstrained variables or arbitrary constraints). Let `(x_q, y_q) = (x_p, -y_p)` and `P != -P` (i.e. `y_p != 0`):
```
(x_a + x_p + x_p) * (x_p - x_p)^2 - (y_p + y_p)^2 = 0
(x_a + 2 x_p) * 0 - 4 y_p^2 = 0 <-- x_a is unconstrained, as expected
                  - 4 y_p^2 = 0
                      y_p   = 0 <-- contradiction
```
The other constraint is fine in this case:
```
(y_a - y_p)(x_p - x_p) - (y_p + y_p)(x_p - x_a) = 0
(y_a - y_p) * 0 - 2 y_p (x_p - x_a) = 0 <-- y_a is unconstrained, as expected
                - 2 y_p (x_p - x_a) = 0
                         x_p = x_a <-- x_a is constrained, but that's fine
```

Reply:
I will clarify this in the spec.